### DISCUSSION TITLE
Experimental Thonny.org py5 sketch runner plug-in

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#32] [Ideas]
CC @tabreturn ... I have no idea of what I'm doing, but... seems to work on my computer.

https://github.com/villares/thonny-py5-runner/

![py5](https://user-images.githubusercontent.com/3694604/127754556-d1ec817a-67d7-4b41-b501-a2057e12ad85.gif)

- Will it work on anyone else's computer?
   -  Looks like this might install another copy of py5 outside the original env, and that's ugly
   -  Can we make the plug-in configure the JAVA_HOME variable? I have no Idea.
- I don't know anything about PyPI
<strike>- I think I should suppress the "on close" dialog if the message is empty</strike> done
- Should I be doing this?
UPDATED: removed the empy message dialog, but... serious questions remain about tapping into the full Thonny's Python debug features and *output for the `print()` function*

**Comment by hx2A** [comment]
> Should I be doing this?

Yes! This is great! Thanks for getting the ball rolling on this.

> Can we make the plug-in configure the JAVA_HOME variable? I have no Idea.

You can write Python code to set environment variables. Something like

```
import os

os.environ['JAVA_HOME'] = <path>
```

The tricky part is writing the code to figure out the correct path to put in there. The user can have more than one java installation on their computer. 

Thonny has some kind of extra help messages when errors are thrown, right? Or is that mu? But in any case, py5 supports customizing all of the error messages. If you wanted to augment the error messages with helpful and targeted advice for how to respond to each error message, this package you are making could add those customizations.

> I don't know anything about PyPI

It's a bit odd but once you get the hang of it you can make and deploy packages with just a few lines of code. When you get to that point we can provide some assistance there. There's also a test environment for you to play around with.

**Reply by hx2A** [reply]
Oh, I looked at the code and saw this: `python_requires='>=3.7',`, but py5 requires a minimum of Python 3.8.

**Reply by tabreturn** [reply]
> Oh, I looked at the code and saw this: `python_requires='>=3.7',`, but py5 requires a minimum of Python 3.8.

Thonny v3.3.7 includes *-alt* versions bundled with Python 3.9.5 (https://github.com/thonny/thonny/releases/tag/v3.3.7). I'm using v3.3.7 until the time the newest release switches to Python 3.8+

**Reply by villares** [reply]
I understood "main" 3.3.7 is already Python 3.8...  (but perhaps I'm interpreting it wrong)

**Reply by tabreturn** [reply]
> I understood "main" 3.3.7 is already Python 3.8

I'm on a Linux machine right now. From what I can tell --

*Windows*
* thonny-3.3.7-windows-portable.zip -- this is Python 3.7 (it contains a python37.dll)
* thonny-3.3.7-py38.exe -- I'm assuming this is Python 3.8?

I couldn't extract the Mac archives properly to see what's in there.

For *Linux*, I can confirm that:

* thonny-3.3.7-x86_64.tar.gz -- this is Python 3.7
* thonny-3.3.7-x86_64-alt.targ.gz -- this is Python 3.9
* thonny-xxl-3.3.7-x86_64.tar.gz -- this is Python 3.7

**Reply by villares** [reply]
Do you think it would it be too bad if the plug-in install script could, asking for permission or warning the user, changed the Thonny config file to point to the correct interpreter (it would have been installed by the py5 installing procedure, done beforehand)?

Using the plug-in a bit and I'm already somewhat frustrated with some truncated error messages :( 

UPDATE... I think the best would be to have rhe person conda install py5, then we'd be certain to have the correct Java 11 etc... maybe miniconda -> conda install py5-thonny bundle....

**Reply by tabreturn** [reply]
> changed the Thonny config file to point to the correct interpreter

I'm guessing it won't be long until Thonny ships with Python 3.9, and people could use the "-alt" version so long. 

> maybe miniconda -> conda install py5-thonny bundle

I'll have to test this out (I haven't tried Conda yet)

**Reply by villares** [reply]
![image](https://user-images.githubusercontent.com/3694604/127899189-0ecdbedb-4563-40d1-89d6-e1a7d99ea469.png)

**Reply by hx2A** [reply]
Good job with the Thonny fact-finding. :)

I do dev work in 3.8 but py5 should work just fine in 3.9 and 3.10. Google Colab uses 3.7, and I want to wait until it catches up.

**Comment by villares** [comment]
I have made another experimental Thonny plug-in, almost exactly like this one, for `shoebot` and I have convinced the maintainer to have a look (and passed the repo to their org). So we will benefit from their insights later on too :D

**Reply by hx2A** [reply]
@villares you're on fire today! And that's collaboration at its finest.

**Comment by tabreturn** [comment]
@villares -- you've inspired me to take a closer look at this :smiley: 

I've experimented with something that'll automatically download and set up JDK and py5:  
https://github.com/tabreturn/thonny-py5mode

This plugin downloads JDK and installs it in the Thonny (portable version) *lib* directory, then sets the `JAVA_HOME` path to that location. You can scan the [__init__.py](https://github.com/tabreturn/thonny-py5mode/blob/main/thonnycontrib/thonny-py5mode/__init__.py) to get the gist of what it does.

**Reply by hx2A** [reply]
@tabreturn I didn't know that https://pypi.org/project/install-jdk/ existed. Thonny-py5mode is a great application of that.

I have finished everything I want to do for the next release and have pushed the code to the 0.5a1.dev0 branch of the py5 repo. I will do some testing and will do a release tomorrow or the next day with that code.

I saw this in the `setup.py`:

```
dependency_links=['https://github.com/hx2A/py5.git@0.5a1.dev0#egg=py5'],
```

Right now that dependency link will get you the code that is about to be released, but after I do the release, I'll delete that branch, so that won't work. I think it is useful to have a branch like that so I can easily share development versions. Otherwise people have to go through all the py5generator stuff, and that might be too much of a hassle.

**Reply by tabreturn** [reply]
Noted, @hx2A. Thanks!

**Reply by villares** [reply]
Wonderful stuff!

You have linked to the thonny-blue-code-format experimet repo, there is also https://github.com/villares/thonny-py5-runner experiment... I'll see if it works with the portable Thonny...

**Reply by tabreturn** [reply]
Doh! I've fixed the repo link.
